Keep package structure in `rootTreeOrProvider`

Duhemm opened the pull request:
`rootTreeOrProvider` is a property of `ClassSymbol`, and contain the
tree that defines this symbol (or a mean to get that tree).

When the tree came from unpickled TASTY files, we were able to see the
nested packages from which the tree comes, along with the imports.
However, when the symbol has been loaded from source, we were filling
the `rootTreeOrProvider` only with the `TypeDef` that introduced this
symbol.

This commit moves the assignment of `rootTreeOrProvider` out of the
typer and into its own phase. Before assigning, we recreate the package
structure so that the information that we get out of
`rootTreeOrProvider` is the same regardless of whether the symbol has
been loaded from source or unpickled from TASTY.

nicolasstucki commented on compiler/src/dotty/tools/dotc/Compiler.scala:
```diff
@@ -52,6 +52,7 @@ class Compiler {
     List(new FirstTransform,         // Some transformations to put trees into a canonical form
          new CheckReentrant,         // Internal use only: Check that compiled program has no data races involving global vars
          new ElimPackagePrefixes,    // Eliminate references to package prefixes in Select nodes
+         new SetRootTree,            // Set the `rootTreeOrProvider` on class symbols
```
When unpickling a symbol, the `rootTreeOrProvider` contains the tree with as it was pickled. Here you are registering the tree after transformations have happened. This may lead us into trouble because `rootTreeOrProvider` will return different trees depending of how we got it.

nicolasstucki replied:
It would make more seance to have the trees that are after PostTyper has transformed them.

nicolasstucki commented on compiler/src/dotty/tools/dotc/transform/SetRootTree.scala:
```diff
@@ -0,0 +1,45 @@
+import dotty.tools.dotc.transform.MegaPhase.MiniPhase
+
+/** Set the `rootTreeOrProvider` property of class symbols. */
+class SetRootTree extends MiniPhase {
```
To traverse top down you should use a `MacroTransform` instead of `MiniPhase`. Look at `PostTyper`

https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/PostTyper.scala#L63-L64

Duhemm replied:
I wanted this to be a mini phase to avoid introducing a whole new traversal, but it seems unavoidable given that this phase has to run right before pickling (there are no miniphases in `frontendPhases`).

This phase doesn't do any transformation of the trees, so it looks like it could be a simple `Phase`, or am I missing something?

nicolasstucki replied:
As I said below, this is not a full traversal and hence it should be quite fast. Linear with the depth of the package assuming you do a top-down traversal and stop when you find the top level class definitions.

Duhemm commented:
test performance please

allanrenucci commented:
Can you clarify what problem this PR solves? What is `rootTreeOrPeovider` used for?

Is there another way to solve this rather than introducing a complete new tree traversal?

nicolasstucki commented:
That would not be a complete tree traversal, it would be quite shallow. Traverse from the root package until you find the first `TypeDef`. The implementation is a bit awkward and makes it look like a full traversal.

allanrenucci commented:
It looks like a complete tree traversal to me:
```scala
    override def traverse(tree: tpd.Tree)(implicit ctx: Context): Unit = {
      if (tree.isInstanceOf[tpd.TypeDef] && tree.symbol.isClass) {
        ...
      }
      traverseChildren(tree)
    }
```
`traverseChildren` is called unconditionally. Am I missing something?

Duhemm commented:
`rootTreeOrProvider` is used to get the tree that introduced a given `ClassSymbol`. This tree can come either from source (for instance, if we have just compiled the file that defined it, or the file that defines it needs recompilation), or it can come from TASTy (we're unpickling).

Given a class symbol, the value of `rootTreeOrProvider` can come from 2 different paths:
 - When the symbol comes from the source, `rootTreeOrProvider` used to be populated by the Typer, where it was set to the typed `TypeDef` that introduced the symbol. We have no information about the packages the `TypeDef` was nested in, nor about the imports.
 - When the symbol comes from TASTy, `rootTreeOrProvider` is set to the unpickler by the symbol loader. When the tree is requested, the unpickler will unpickle the trees, which will include the nested package defs and imports.

In the IDE, we always want to see the top level imports, because this information is necessary to correctly do renamings.
With the IDE, we have 3 scenarios to consider:
 1. The file is open in the IDE: We will have the complete tree without needing to look at `rootTreeOrProvider`, so we'll see the complete structure. 
 2. The file is not open in the IDE, and it doesn't need recompilation: We will unpickle the tree and see the complete structure.
 3. The file is not open in the IDE, and it needs recompilation: We will use `rootTreeOrProvider`, and we won't be able to see top level imports.

This PR is only important in the 3rd scenario. When setting the `rootTreeOrProvider`, we take the typed compilation unit, keep the nested packages and imports and the class definition. This way, we can store in `rootTreeOrProvider` an information similar to what we'd if we were unpickling the tree. Doing this is moved out of the typer so that we have access to the typed compilation unit.

Now that we have IDE tests where the files are not "open in the IDE", I'll add tests where that matters.

allanrenucci commented:
Also I don't think the benchmark set `YretainTrees`, so you would not be able to measure the impact of this change

nicolasstucki commented:
> Also I don't think the benchmark set YretainTrees, so you would not be able to measure the impact of this change

No, they do not

Duhemm commented on compiler/src/dotty/tools/dotc/transform/SetRootTree.scala:
```diff
@@ -3,41 +3,34 @@ package dotty.tools.dotc.transform
+            sym.rootTreeOrProvider = tree
+        }
+      }
+      traverseChildren(tree)
```
Doesn't that mean that we would not be able to find the tree for an inner class directly?

With current master, for an inner class, we'd get the `TypeDef` if the source file needs to be `lateCompile`d, or the `PackageDef` otherwise.

With the change you suggest, we would always have an empty tree for inner classes. One would need to get the tree of the enclosing top level class and navigate from there.

Looking at `rootTreeOrProvider`, `rootTree` and `rootTreeContaining`, I'm under the impression that only `SourceTree` and `ReadTastyFromClasses` may need to be adapted. Can you think of something else that'd require changes?

Duhemm replied:
Actually, it looks like `rootTree` is only supposed to be called on top level class symbols, so everything should be fine, thanks.

Duhemm commented on compiler/src/dotty/tools/dotc/Run.scala:
```diff
@@ -215,7 +215,7 @@ class Run(comp: Compiler, ictx: Context) extends ImplicitRunInfo with Constraint
           unit.tpdTree = ctx.typer.typedExpr(unit.untpdTree)
           val phase = new transform.SetRootTree()
-          phase.runOn(unit)
+          phase.run
```
`runOn` was a method I'd defined in `SetRootTree`, but it was no longer after I made `SetRootTree` a `Phase`.

Duhemm commented:
@nicolasstucki I've added a test that illustrates why this patch is needed. Without the addition of `SetRootTree`, the test fails as expected.

I've addressed your comments:
 - The traversal is now very shallow
 - I've moved from `Phase` to `MacroTransform`. I'm not convinced this is an improvement, since this phase doesn't change the tree.

allanrenucci commented:
I think 6272e8d should be reverted. If it does not transform the tree then it doesn't have to be a macro transform

nicolasstucki commented:
@Duhemm  something failed in the tests

nicolasstucki commented on compiler/src/dotty/tools/dotc/transform/SetRootTree.scala:
```diff
@@ -0,0 +1,41 @@
+              val sym = td.symbol.asClass
+              tpd.sliceTopLevel(ctx.compilationUnit.tpdTree, sym) match {
+                case (pkg: tpd.PackageDef) :: Nil =>
+                  sym.rootTreeOrProvider = pkg
```
The `rootTreeOrProvider` should only be set in the top level class

Duhemm replied:
I think it's the case already since this no longer traverses the body of classes, no?

nicolasstucki replied:
I misinterpreted the code. It is fine.

nicolasstucki commented:
> I've moved from Phase to MacroTransform. I'm not convinced this is an improvement, since this phase doesn't change the tree.

Yes you are right, it is better as a phase.

Duhemm commented:
@nicolasstucki I've fixed the test in 136b0ceb4ff6419372cc94a1fb82a43c75048cc7, do you mind taking another look?
